File: problems/problem_004/python/solution.py
def zigzag_conversion(s: str, num_rows: int) -> str:
    if num_rows == 1:
        # no zigzag string possible
        return s

    result = [''] * num_rows

    current_row = 0
    traverse_down = False

    for char in s:
        result[current_row] += char

        if current_row == 0 or current_row == num_rows - 1:
            traverse_down = not traverse_down

        current_row += 1 if traverse_down else -1

    return ''.join(result)

File: problems/problem_004/javascript/solution.js
export function zigzagConversion(s, numRows) {
    if (numRows === 1) return s;

    const result = new Array(numRows).fill('');
    let currentRow = 0;
    let traverseDown = false;

    for (let char of s) {
        result[currentRow] += char;

        if (currentRow === 0 || currentRow === numRows - 1) {
            traverseDown = !traverseDown;
        }

        // move to next row based on direction
        currentRow += traverseDown ? 1 : -1;
    }

    return result.join('');
}
